### ISSUE TITLE
v2.0.0 chokes on PostgreSQL's `SELECT DISTINCT ON`

### ISSUE THREAD

**Issue by epgui** [sqlfluff/sqlfluff#4536]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

If I try to lint this:

```sql
SELECT DISTINCT ON (property_name)
    property_name,
    NULL AS raw_value,
    NULL AS "value",
    -999 AS member_id
FROM my_table
```

Then I get something like:

```
L:  1 | P:  38 | LT02 | Expected line break and no indent before ')'.
                      | [layout.indent]
L:  2 | P:   1 | LT02 | Expected indent of 12 spaces.
                      | [layout.indent]
L:  3 | P:   1 | LT02 | Expected indent of 12 spaces.
                      | [layout.indent]
L:  4 | P:   1 | LT02 | Expected indent of 12 spaces.
                      | [layout.indent]
L:  5 | P:   1 | LT02 | Expected indent of 12 spaces.
                      | [layout.indent]
L:  6 | P:   1 | LT02 | Expected indent of 8 spaces.
                      | [layout.indent]
L:  7 | P:   1 | LT02 | Expected indent of 8 spaces.
                      | [layout.indent]
```

And if I try and let sqlfluff "fix" it, I get this:

```sql
SELECT DISTINCT ON (property_name
)
    property_name,
    NULL AS raw_value,
    NULL AS "value",
    -999 AS member_id
FROM my_table
```

### Expected Behaviour

I believe the provided snippet should simply pass.

### Observed Behaviour

The provided snippet fails.

### How to reproduce

This example lints just fine:

```sql
SELECT DISTINCT ON (
    my_table.member_id,
    my_table.property_name
)
    my_table.member_id,
    my_table.property_name,
    my_table.raw_value,
    my_table."value"
FROM ids
LEFT JOIN my_table
    ON ids.member_id = my_table.member_id
```

This example doesn't lint properly:

```sql
SELECT DISTINCT ON (
    property_name
)
    property_name,
    NULL AS raw_value,
    NULL AS "value",
    -999 AS member_id
FROM my_table
```

And neither does this one:

```sql
SELECT DISTINCT ON (property_name)
    property_name,
    NULL AS raw_value,
    NULL AS "value",
    -999 AS member_id
FROM my_table
```

But if I do this, then the only rule that fails is `references.consistent`:

```sql
SELECT DISTINCT ON (my_table.property_name)
    property_name,
    NULL AS raw_value,
    NULL AS "value",
    -999 AS member_id
FROM my_table
```

### Dialect

PostgreSQL

### Version

2.0.0

### Configuration

```ini
[sqlfluff]
dialect = postgres
; [ST07][structure.using]: Prefer specifying join keys instead of using USING.
; This rule was originally intended to encourage better intercompatibility
; across databases, but it would appear that whatever issues existed with USING
; statements no longer exist, and that some would now encourage their use.
; See also:
;   - https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.rules.Rule_L032
;   - https://github.com/dbt-labs/corp/pull/58
exclude_rules = structure.using
max_line_length = 100
; Use all available CPU processes when linting
processes = 0
templater = jinja

[sqlfluff:indentation]
indent_unit = space
indented_using_on = true
tab_space_size = 4

[sqlfluff:layout:type:binary_operator]
line_position = trailing

[sqlfluff:rules:aliasing.table]
aliasing = explicit

[sqlfluff:rules:aliasing.column]
aliasing = explicit

[sqlfluff:rules:aliasing.expression]
allow_scalar = false

[sqlfluff:rules:aliasing.length]
; PostgreSQL NAMEDATALEN is 63 bytes
max_alias_length = 63
min_alias_length = 1

[sqlfluff:rules:aliasing.forbid]
force_enable = false

[sqlfluff:rules:ambiguous.join]
fully_qualify_join_types = inner

[sqlfluff:rules:ambiguous.column_references]
group_by_and_order_by_style = explicit

[sqlfluff:rules:capitalisation.keywords]
capitalisation_policy = upper

[sqlfluff:rules:capitalisation.identifiers]
extended_capitalisation_policy = lower
unquoted_identifiers_policy = all

[sqlfluff:rules:capitalisation.functions]
extended_capitalisation_policy = upper

[sqlfluff:rules:capitalisation.literals]
extended_capitalisation_policy = upper

[sqlfluff:rules:capitalisation.types]
extended_capitalisation_policy = upper

[sqlfluff:rules:convention.select_trailing_comma]
select_clause_trailing_comma = forbid

[sqlfluff:rules:convention.count_rows]
prefer_count_0 = false
prefer_count_1 = false

[sqlfluff:rules:convention.terminator]
multiline_newline = false
require_final_semicolon = true

[sqlfluff:rules:convention.quoted_literals]
force_enable = true
preferred_quoted_literal_style = single_quotes

[sqlfluff:rules:convention.casting_style]
preferred_type_casting_style = consistent

[sqlfluff:rules:layout.long_lines]
ignore_comment_clauses = false
ignore_comment_lines = false

[sqlfluff:rules:layout.select_targets]
wildcard_policy = single

[sqlfluff:rules:references.from]
force_enable = true

[sqlfluff:rules:references.consistent]
force_enable = true
single_table_references = unqualified

[sqlfluff:rules:references.keywords]
quoted_identifiers_policy = none
unquoted_identifiers_policy = all

[sqlfluff:rules:references.special_chars]
allow_space_in_identifier = false
quoted_identifiers_policy = none
unquoted_identifiers_policy = all

; [sqlfluff:rules:references.quoting]
; force_enable = true
; prefer_quoted_identifiers = false

[sqlfluff:rules:structure.subquery]
forbid_subquery_in = both

[sqlfluff:templater]
unwrap_wrapped_queries = true

[sqlfluff:templater:jinja]
apply_dbt_builtins = true
library_path = sqlfluff_libs
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
